src/main/java/org/forgerock/openam/auth/nodes/SetProfilePropertyNode.java

```diff
@@ -90,41 +112,63 @@ public SetProfilePropertyNode(@Assisted Config config, CoreWrapper coreWrapper)
-            String propertyValue = entry.getValue();
-
-            String result = null;
-            if (propertyValue.startsWith("\"")) {
-                result = propertyValue.substring(1, propertyValue.length() - 1);
-            } else if (context.sharedState.isDefined(propertyValue)) {
-                result = context.sharedState.get(propertyValue).asString();
-            }
-
-            if (StringUtils.isNotEmpty(result)) {
-                attributes.put(key, singleton(result));
-            }
-        }
-
-        config.transientProperties().forEach((key, value) -> {
-            if (context.transientState.isDefined(value)) {
-                attributes.put(key, singleton(context.transientState.get(value).asString()));
-            }
-        });
+        Map<String, Set<String>> attributes = createAttributes(context.transientState, userIdentity,
+                                                               config.transientProperties(),
+                                                               createAttributes(context.sharedState, userIdentity,
+                                                                                config.properties(), new HashMap<>()));
 
         try {
             userIdentity.setAttributes(attributes);
             userIdentity.store();
         } catch (IdRepoException | SSOException ex) {
-            LOGGER.error("Unable to update user {} in realm {} with attributes {}", username, realm, attributes, ex);
+            LOGGER.error("Unable to update user " + username + " in realm " + realm + " with attributes " + attributes,
+                         ex);
         }
 
         return goToNext().build();
     }
+
+    private Map<String, Set<String>> createAttributes(JsonValue state, AMIdentity userIdentity,
+                                                      Map<String, String> properties,
+                                                      Map<String, Set<String>> attributes)
+            throws NodeProcessException {
+        for (Map.Entry<String, String> entry : properties.entrySet()) {
+            String key = entry.getKey();
+            String propertyValue = entry.getValue();
+
+            Set<String> result = null;
+            if (propertyValue.startsWith("\"")) {
+                result = Collections.singleton(propertyValue.substring(1, propertyValue.length() - 1));
+            } else if (state.isDefined(propertyValue)) {
+                if (config.isMultiValuedString()) {
+                    result = new HashSet<>(state.get(propertyValue).asList(String.class));
+                } else {
+                    result = Collections.singleton(state.get(propertyValue).asString());
+                }
+            }
+
+            if (!result.isEmpty()) {
+                if (config.addAttributes()) {
+                    Set<String> oldAttributes;
+                    try {
+                        oldAttributes = userIdentity.getAttribute(key);
```

Comment:
Each getAttribute invocation can result in an individual LDAP call. Preferably we should use getAttributes with all the relevant attributes at the end of the processing and then update the attribute values based on that one call.